z3/src/solver.rs

```diff
@@ -268,13 +268,13 @@ impl<'ctx> Solver<'ctx> {
     // have an Ast subtype for yet.
     pub fn get_proof(&self) -> Option<impl Ast<'ctx>> {
         let guard = Z3_MUTEX.lock().unwrap();
-        Some(ast::Dynamic::new(self.ctx, unsafe {
-            let m = Z3_solver_get_proof(self.ctx.z3_ctx, self.z3_slv);
-            if m.is_null() {
-                return None;
-            }
-            m
-        }))
+        let m = unsafe { Z3_solver_get_proof(self.ctx.z3_ctx, self.z3_slv) };
+        std::mem::drop(guard);
```

Comment:
Can you make this into

```rust
let m = unsafe {
    let _guard = Z3_MUTEX.lock().unwrap();
    Z3_solver_get_proof(...)
};
```

Reply:
Can Rust delay dropping a structure if optimizations are on? I was trying to find clarification on the rules, so I was explicit, but if implicit works too I can totally change it.

Reply:
No, I just find it easier to read the code and follow what is going on with explicit blocks.

Reply:
Fixed!